src/reducers/scrollHelper.js

```diff
@@ -0,0 +1,459 @@
+    rows[rowPosition] = rowIndex;
+  } else {
+    // This row already is in the table with rowPosition position or it
+    // can replace row that is in that position
+    rows[rowPosition] = rowIndex;
+  }
+};
+
+/**
+ * @param {!Object} state
+ * @private
+ */
+function _updateRowsWithBuffer(state) {
+  let { bufferRowsCount, viewportRowsBegin, viewportRowsEnd, rowsCount } = state;
+  let remainingBufferRows = 2 * bufferRowsCount;
+  let bufferRowIndex = Math.max(viewportRowsBegin - bufferRowsCount, 0);
+  while (bufferRowIndex < viewportRowsBegin) {
+    _addRowToBuffer(
+      state,
+      bufferRowIndex,
+      viewportRowsBegin,
+      viewportRowsEnd - 1
+    );
+    bufferRowIndex++;
+    remainingBufferRows--;
+  }
+  bufferRowIndex = viewportRowsEnd;
+  while (bufferRowIndex < rowsCount && remainingBufferRows > 0) {
+    _addRowToBuffer(
+      state,
+      bufferRowIndex,
+      viewportRowsBegin,
+      viewportRowsEnd - 1
+    );
+    bufferRowIndex++;
+    remainingBufferRows--;
+  }
+}
+
+/**
+ * @param {!Object} state
+ * @private
+ */
+function _updateRows(state) {
+  let { firstRowIndex, firstRowOffset, maxVisibleRowCount, rowsCount, viewportHeight, rowHeightGetter } = state;
+
+  let top = firstRowOffset;
+  let totalHeight = top;
+  let rowIndex = firstRowIndex;
+  let endIndex = Math.min(firstRowIndex + maxVisibleRowCount, rowsCount);
+
+  state.viewportRowsBegin = firstRowIndex;
+  while (rowIndex < endIndex || (totalHeight < viewportHeight && rowIndex < rowsCount)) {
+    _addRowToBuffer(
+      state,
+      rowIndex,
+      firstRowIndex,
+      endIndex - 1
+    );
+    totalHeight += rowHeightGetter(rowIndex);
+    ++rowIndex;
+    // Store index after the last viewport row as end, to be able to
+    // distinguish when there are no rows rendered in viewport
+    state.viewportRowsEnd = rowIndex;
+  }
+};
+
+/**
+ * @param {!Object} state
+ * @private
+ */
+function _recalculateRowHeights(state) {
+  let { rows, rowHeights, rowOffsets } = state;
+  state.rowHeights = {};
+
+  //Sort the rows, we slice first to avoid changing original
+  let sortedRowsToRender = rows.slice().sort((a, b) => a - b);
+  sortedRowsToRender.forEach((rowIndex) => {
+    _updateRowHeight(state, rowIndex);
+    state.rowHeights[rowIndex] = rowOffsets.sumUntil(rowIndex);
+  });
+}
+
+/**
+ * Creates new state object when rowHeight or rowCount changes
+ *
+ * @param {!Object} state
+ * @param {!Object} props
+ * @return {!Object}
+ */
+function initailize(state, props) {
+  var viewportHeight = _caculateViewportHeight(props);
+  let { rowsCount, rowHeight, rowHeightGetter } = props;
+  let maxVisibleRowCount = Math.ceil(viewportHeight / rowHeight) + 1;
+  let bufferRowsCount = clamp(
+    Math.floor(maxVisibleRowCount / 2),
+    MIN_BUFFER_ROWS,
+    MAX_BUFFER_ROWS
+  );
+
+  let storedHeights = new Array(rowsCount);
+  for (var i = 0; i < rowsCount; i++) {
+    storedHeights[i] = rowHeight;
+  }
+
+  state = Object.assign({}, state, {
```

Comment:
Is there any documentation on what all these fields are and how they are / can be updated / changed?  It's currently a challenge to follow the logic here.